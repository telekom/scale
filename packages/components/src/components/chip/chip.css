/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host {
  --height: 32px;

  --font-size: var(--telekom-typography-font-size-body);
  --line-height: var(--telekom-typography-line-spacing-standard);
  --font-weight: var(--telekom-typography-font-weight-bold);
  --color-focus: var(--telekom-color-functional-focus-standard);
  --focus-outline: var(--telekom-line-weight-highlight) solid
    var(--telekom-color-functional-focus-standard);
  --box-shadow: 0px 1px 2px rgba(0, 0, 0, 0.1), 0px 2px 8px rgba(0, 0, 0, 0.1);

  /*standard selected*/
  --background-standard-selected: var(--telekom-color-ui-extra-strong);
  --color-standard-selected: var(
    --telekom-color-text-and-icon-inverted-standard
  );
}
/**TODO: colors unexplained in figma  **/
.chip {
  box-sizing: border-box;
  border: var(--telekom-spacing-composition-space-01) solid transparent;
  display: inline-flex;
  outline: none;
  padding: 0 var(--telekom-spacing-composition-space-05);
  text-align: center;
  align-items: center;
  white-space: nowrap;
  border-radius: 1rem;
  border-color: rgba(255, 255, 255, 0.14);
  height: var(--height);
  vertical-align: baseline;
  justify-content: center;
  cursor: default;
  color: var(--telekom-color-text-and-icon-standard);
  background: var(--telekom-color-text-and-icon-inverted-standard);
  border: 1px solid var(--telekom-color-ui-border-standard);
}

.chip.chip--selected.chip--type-dynamic {
  padding-right: var(--telekom-spacing-composition-space-03);
}

.chip.chip--selected.chip--type-dynamic button {
  border: none;
  cursor: pointer;
  padding: 0;
  height: 24px;
  width: 24px;
  outline: none;
  background: transparent;
  border-radius: var(--telekom-radius-circle);
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: transparent;
}

.chip.chip--selected.chip--type-dynamic.chip--variant-standard button:hover {
  background: var(--telekom-color-ui-state-fill-hovered-inverted);
}

.chip.chip--selected.chip--type-dynamic.chip--variant-standard button:active {
  background: var(--telekom-color-ui-state-fill-pressed-inverted);
}

.chip.chip--selected.chip--type-dynamic button:focus {
  justify-content: center;
  outline: var(--focus-outline);
}

.chip.chip--selected.chip--type-dynamic.chip--variant-standard
  scale-icon-action-close {
  padding: 0;
  color: var(--telekom-color-text-and-icon-inverted-standard);
}

.chip:not(.chip--disabled):not(.chip--type-dynamic):focus,
.chip.chip--type-dynamic:not(.chip--selected):focus {
  outline: var(--telekom-spacing-composition-space-02) solid var(--color-focus);
  outline-offset: var(--telekom-spacing-composition-space-01);
}

.chip:not(.chip--disabled):not(.chip--type-dynamic):hover,
.chip.chip--type-dynamic:not(.chip--selected):hover {
  background: linear-gradient(
      var(--telekom-color-ui-state-fill-hovered),
      var(--telekom-color-ui-state-fill-hovered)
    ),
    var(--telekom-color-text-and-icon-inverted-standard);
}

.chip:not(.chip--disabled):not(.chip--type-dynamic):active,
.chip.chip--type-dynamic:not(.chip--selected):active {
  background: linear-gradient(
      var(--telekom-color-ui-state-fill-pressed),
      var(--telekom-color-ui-state-fill-pressed)
    ),
    var(--telekom-color-text-and-icon-inverted-standard);
}

.chip--selected {
  background: var(--background-standard-selected);
  color: var(--color-standard-selected);
}

.chip--selected:not(.chip--disabled):not(.chip--type-dynamic):hover {
  background: linear-gradient(
      var(--telekom-color-ui-state-fill-hovered-inverted),
      var(--telekom-color-ui-state-fill-hovered-inverted)
    ),
    var(--background-standard-selected);
}

.chip--selected:not(.chip--disabled):not(.chip--type-dynamic):active {
  background: linear-gradient(
      var(--telekom-color-ui-state-fill-pressed-inverted),
      var(--telekom-color-ui-state-fill-pressed-inverted)
    ),
    var(--background-standard-selected);
}

/**
label
**/

.chip-label {
  padding: 0 var(--telekom-spacing-composition-space-03) 0
    var(--telekom-spacing-composition-space-03);
  font-weight: var(--telekom-typography-font-weight-bold);
  font-size: 1rem;
  line-height: 100%;
}

.chip--selected .chip-label {
  padding-right: var(--telekom-spacing-composition-space-04);
}
.chip--type-dynamic.chip--selected .chip-label {
  padding-right: var(--telekom-spacing-composition-space-03);
}

/**
icon front (slot)
**/

.chip slot[name='chip-icon']::slotted(*) {
  padding-right: var(--telekom-spacing-composition-space-03);
  padding-top: var(--telekom-spacing-composition-space-03);
}

/**
icon end
*/

/**
disabled
**/
.chip.chip--disabled:not(.chip--type-dynamic) {
  color: var(--telekom-color-text-and-icon-disabled);
  border: 1px solid var(--telekom-color-ui-border-disabled);
}

.chip--selected:not(.chip--variant-outlined):not(.chip--type-dynamic).chip--disabled {
  background: var(--telekom-color-ui-disabled);
  color: var(--telekom-color-text-and-icon-disabled);
  border: none;
}

.chip--variant-outlined.chip--selected.chip--disabled:not(.chip--type-dynamic) {
  background-color: var(--telekom-color-text-and-icon-inverted-standard);
  color: var(--telekom-color-text-and-icon-disabled);
  border: 1px solid var(--telekom-color-ui-border-disabled);
}

/**
variant outline
**/

.chip.chip--variant-outline.chip--selected:not(.chip.chip--disabled):not(.chip--type-dynamic):hover {
  background: linear-gradient(
      var(--telekom-color-ui-state-fill-hovered),
      var(--telekom-color-ui-state-fill-hovered)
    ),
    var(--telekom-color-text-and-icon-inverted-standard);
}

.chip.chip--variant-outline.chip--selected:not(.chip.chip--disabled):not(.chip--type-dynamic):active {
  background: linear-gradient(
      var(--telekom-color-ui-state-fill-pressed),
      var(--telekom-color-ui-state-fill-pressed)
    ),
    var(--telekom-color-text-and-icon-inverted-standard);
}

/**
dynamic
**/

.chip.chip--variant-outline.chip--selected {
  color: var(--telekom-color-text-and-icon-primary-standard);
  background: var(--telekom-color-text-and-icon-inverted-standard);
  border: 1px solid var(--telekom-color-text-and-icon-primary-standard);
}

.chip.chip--selected.chip--type-dynamic.chip--variant-outline
  scale-icon-action-close {
  padding: 0;
  color: var(--telekom-color-text-and-icon-primary-standard);
}

.chip.chip--selected.chip--type-dynamic.chip--variant-outline button:hover {
  background: var(--telekom-color-ui-state-fill-hovered);
}

.chip.chip--selected.chip--type-dynamic.chip--variant-outline
  button:hover
  scale-icon-action-close {
  color: var(--telekom-color-text-and-icon-primary-hovered);
}

.chip.chip--selected.chip--type-dynamic.chip--variant-outline button:active {
  background: var(telekom-color-ui-state-fill-pressed);
}

.chip.chip--selected.chip--type-dynamic.chip--variant-outline
  button:active
  scale-icon-action-close {
  color: var(--telekom-color-text-and-icon-primary-pressed);
}
