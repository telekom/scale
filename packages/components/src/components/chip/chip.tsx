/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import {
  Component,
  Prop,
  h,
  Host,
  Event,
  EventEmitter,
  Element,
} from '@stencil/core';
import classNames from 'classnames';
import { emitEvent } from '../../utils/utils';
@Component({
  tag: 'scale-chip',
  styleUrl: './chip.css',
  shadow: true,
})
export class Chip {
  @Element() hostElement: HTMLElement;
  /** (optional) */
  @Prop() variant?: 'standard' | 'outline' = 'standard';
  /** (optional) */
  @Prop() type?: 'dynamic' | 'persistent' = 'persistent';
  /** (optional) */
  @Prop() selected?: boolean = false;
  /** (optional) chip aria-role */
  @Prop() ariaRoleTitle?:
    | 'switch'
    | 'radio'
    | 'option'
    | 'menuitemreadio'
    | 'menuitemcheckbox'
    | 'checkbox' = 'switch';
  /** @deprecated (optional) chip aria-checked - should be derived from selected state attribute */
  @Prop() ariaCheckedState?: boolean = false;
  /** (optional) chip label */
  @Prop() label?: string;
  /** (optional) chip disabled */
  @Prop() disabled?: boolean = false;
  /** (optional) Dismiss label */
  @Prop() dismissText?: string = 'dismiss';
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;

  /** (optional) Change icon click event */
  @Event({ eventName: 'scale-change' }) scaleChange: EventEmitter<MouseEvent>;
  /** @deprecated in v3 in favor of kebab-case event names */
  @Event({ eventName: 'scaleChange' })
  scaleChangeLegacy: EventEmitter<MouseEvent>;
  /** (optional) Close icon click event */
  @Event({ eventName: 'scale-close' }) scaleClose: EventEmitter<MouseEvent>;
  /** @deprecated in v3 in favor of kebab-case event names */
  @Event({ eventName: 'scaleClose' })
  scaleCloseLegacy: EventEmitter<MouseEvent>;

  componentDidRender() {
    // handle no setted icon size attribute
    const defaultIconSize = 24;
    const iconSlot = this.hostElement.querySelector(
      '[slot="chip-icon"]'
    ) as HTMLElement;
    if (iconSlot !== null) {
      if (
        iconSlot.children[0].getAttribute('size') === String(defaultIconSize)
      ) {
        iconSlot.children[0].setAttribute('size', String(16));
      }
      if (this.selected) {
        iconSlot.children[0].setAttribute('selected', String(true));
      } else {
        iconSlot.children[0].setAttribute('selected', String(false));
      }
    }
  }
  disconnectedCallback() {}

  handleClose = (event: MouseEvent) => {
    event.preventDefault();
    event.stopPropagation();
    if (this.disabled && this.type !== 'dynamic') {
      return;
    }
    emitEvent(this, 'scaleClose', event);
  };

  handleClick = (event: MouseEvent) => {
    this.handleChangeEvent(event);
  };

  handleKeyDown = (event: KeyboardEvent) => {
    if (event.code === 'Space') {
      this.handleChangeEvent(event);
    }
  };

  handleChangeEvent = (event: MouseEvent | KeyboardEvent): void => {
    if (this.disabled && this.type === 'dynamic') {
      return;
    }
    if (this.type !== 'dynamic') {
      this.selected = !this.selected;
    }
    event.preventDefault();
    event.stopPropagation();
    if (this.disabled && this.type !== 'dynamic') {
      return;
    }
    emitEvent(this, 'scaleChange', event);
  };

  getIcon() {
    if (this.type === 'dynamic' && this.selected) {
      return (
        <button
          part="button-dismissable"
          disabled={this.disabled}
          aria-label={this.dismissText}
          onClick={!this.disabled ? this.handleClose : null}
        >
          <scale-icon-action-close
            accessibility-title="close"
            size={16}
            selected
          />
        </button>
      );
    } else if (this.type === 'persistent' && this.selected) {
      return (
        <scale-icon-action-checkmark
          accessibility-title="success"
          size={16}
          selected
        />
      );
    } else if (this.type === 'persistent') {
      return (
        <scale-icon-action-checkmark accessibility-title="success" size={16} />
      );
    }
  }

  render() {
    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}
        <span
          role={this.ariaRoleTitle}
          aria-checked={this.selected.toString()}
          tabindex={this.disabled ? '-1' : '0'}
          part={this.getBasePartMap()}
          class={this.getCssClassMap()}
          onClick={this.handleClick}
          onKeyDown={this.handleKeyDown}
        >
          <slot name="chip-icon"></slot>
          <span class="chip-label">
            <slot />
          </span>
          {this.selected ? this.getIcon() : null}
        </span>
      </Host>
    );
  }

  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }

  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }

  getCssOrBasePartMap(mode: 'basePart' | 'css') {
    const component = 'chip';
    const prefix = mode === 'basePart' ? '' : `${component}--`;

    return classNames(
      mode === 'basePart' ? 'base' : component,
      !!this.selected && `${prefix}selected`,
      !!this.disabled && `${prefix}disabled`,
      this.type && `${prefix}type-${this.type}`,
      this.variant && `${prefix}variant-${this.variant}`
    );
  }
}
