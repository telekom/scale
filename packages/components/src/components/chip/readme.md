# scale-chip



<!-- Auto Generated Below -->


## Properties

| Property           | Attribute            | Description                                                                                                                              | Type                                                                                      | Default        |
| ------------------ | -------------------- | ---------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------- | -------------- |
| `ariaCheckedState` | `aria-checked-state` | <span style="color:red">**[DEPRECATED]**</span> (optional) chip aria-checked - should be derived from selected state attribute<br/><br/> | `boolean`                                                                                 | `false`        |
| `ariaRoleTitle`    | `aria-role-title`    | (optional) chip aria-role                                                                                                                | `"checkbox" \| "menuitemcheckbox" \| "menuitemreadio" \| "option" \| "radio" \| "switch"` | `'switch'`     |
| `disabled`         | `disabled`           | (optional) chip disabled                                                                                                                 | `boolean`                                                                                 | `false`        |
| `dismissText`      | `dismiss-text`       | (optional) Dismiss label                                                                                                                 | `string`                                                                                  | `'dismiss'`    |
| `label`            | `label`              | (optional) chip label                                                                                                                    | `string`                                                                                  | `undefined`    |
| `selected`         | `selected`           | (optional)                                                                                                                               | `boolean`                                                                                 | `false`        |
| `styles`           | `styles`             | (optional) Injected CSS styles                                                                                                           | `string`                                                                                  | `undefined`    |
| `type`             | `type`               | (optional)                                                                                                                               | `"dynamic" \| "persistent"`                                                               | `'persistent'` |
| `variant`          | `variant`            | (optional)                                                                                                                               | `"outline" \| "standard"`                                                                 | `'standard'`   |


## Events

| Event          | Description                                                                                        | Type                      |
| -------------- | -------------------------------------------------------------------------------------------------- | ------------------------- |
| `scale-change` | (optional) Change icon click event                                                                 | `CustomEvent<MouseEvent>` |
| `scale-close`  | (optional) Close icon click event                                                                  | `CustomEvent<MouseEvent>` |
| `scaleChange`  | <span style="color:red">**[DEPRECATED]**</span> in v3 in favor of kebab-case event names<br/><br/> | `CustomEvent<MouseEvent>` |
| `scaleClose`   | <span style="color:red">**[DEPRECATED]**</span> in v3 in favor of kebab-case event names<br/><br/> | `CustomEvent<MouseEvent>` |


## Shadow Parts

| Part                   | Description |
| ---------------------- | ----------- |
| `"button-dismissable"` |             |


## Dependencies

### Depends on

- [scale-icon-action-close](../icons/action-close)
- [scale-icon-action-checkmark](../icons/action-checkmark)

### Graph
```mermaid
graph TD;
  scale-chip --> scale-icon-action-close
  scale-chip --> scale-icon-action-checkmark
  style scale-chip fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
