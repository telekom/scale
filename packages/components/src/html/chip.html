<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0"
    />
    <title>Stencil Component Starter</title>

    <script type="module" src="/build/scale-components.esm.js"></script>
    <link rel="stylesheet" href="/build/scale-components.css" />
    <style type="text/css" media="screen, print">
      html {
        margin: 0;
        padding: 0;
      }

      body {
        margin: 0;
        padding: 4rem;
      }
    </style>
  </head>

  <body>
    <p>
      <scale-chip>Label</scale-chip>
      <scale-chip disabled>Label</scale-chip>
      <scale-chip selected>Label</scale-chip>
      <scale-chip type="dynamic">Label</scale-chip>
      <scale-chip disabled type="dynamic">Label</scale-chip>
      <scale-chip selected type="dynamic">Label</scale-chip>
    </p>
  </body>
</html>
