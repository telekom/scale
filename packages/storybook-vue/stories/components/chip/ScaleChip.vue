<template>
  <scale-chip
    :type="type"
    :variant="variant"
    :selected="selected"
    :aria-checked-state="ariaCheckedState"
    :aria-role-title="ariaRoleTitle"
    :disabled="disabled"
    :styles="styles"
    @scaleChange="scaleChange"
    @scaleClose="scaleClose"
  >
    <slot name="chip-icon"></slot>
    <slot />
  </scale-chip>
</template>

<script>
import { action } from '@storybook/addon-actions';
export default {
  props: {
    type: { type: 'dynamic' | 'persistent', default: 'persistent' },
    variant: { type: 'standard' | 'outline', default: 'standard' },
    selected: { type: Boolean, default: false },
    disabled: { type: Boolean, default: false },
    styles: { type: 'switch' | 'radio' | 'option' | 'menuitemreadio' | 'menuitemcheckbox' | 'checkbox', default: 'switch'},
    ariaRoleTitle: { type: String },
    ariaCheckedState: { type: Boolean },
  },
  methods: {
    scaleChange($event) {
      action('scaleChange');
      this.$emit('scaleChange', $event);
    },
    'scale-change'($event) {
      action('scale-change');
      this.$emit('scale-change', $event);
    },
    scaleClose($event) {
      action('scaleClose');
      this.$emit('scaleClose', $event);
    },
    'scale-close'($event) {
      action('scale-close');
      this.$emit('scale-close', $event);
    },
  },
};
</script>
