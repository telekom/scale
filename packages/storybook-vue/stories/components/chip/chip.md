<div style="display: inline-flex; align-items: center; justify-content: space-between; width: 100%;">
    <h1>Chip</h1>
    <img src="assets/beta.png" alt="Beta Component" />
</div>

Chips are contextual components that allow users to filter content or trigger actions.

## General

Chips consist of relevant keywords in a container. They often appear dynamically (triggered by a user action) and as a group.

### When to use

Use chips to:

- give users filtering possibilities,
- display the filters users set, or
- suggest options to users (search criteria, possible answers).

### When not to use

Don’t use chips when:

- you want to mark a category (use tags instead) or
- you can use a button instead.

## Basic types

![Image Name](assets/3_components/chip/Chips-Types.png)

### Persistent chips

Persistent chips are an integral part of the UI and can be selected and deselected.

![Image Name](assets/3_components/chip/Chips-Persistent-Default.png)

**Example 1 – Default** <br/>
A group of filter options as an integral part of the UI.

![Image Name](assets/3_components/chip/Chips-Persistent-Selected.png)

**Example 1 – Filter selected** <br/>
Users can recognize selected filters with the checkmark icon.

### Dynamic chips

Dynamic chips fade in or out after a user action.

![Image Name](assets/3_components/chip/Chips-Dynamic-Example1.png)

**Example 1** <br/>
The chip appears after users set a filter. Clicking/tapping on the close icon hides it again and resets the filter.

![Image Name](assets/3_components/chip/Chips-Dynamic-Example_Chatbot2.png)

**Example 2** <br/>
A set of possible responses in a chat. After clicking/tapping on the chip, it fades out again.

## Variants

<table> <tr><th>![Image Name](assets/3_components/chip/Chips-Variant-Outline.png)</th><th>![Image Name](assets/3_components/chip/Chips-Variant-Standard.png)</th></tr></table>

Chips can be displayed in two variants: standard and outline.

### Disabled state

<table> <tr><th>![Image Name](assets/3_components/chip/Chips-Disabled-Outline-EN.png)</th><th>![Image Name](assets/3_components/chip/Chips-Disabled-Standard-EN.png)</th></tr></table>

In the disabled state, the chip's interactive function is switched off. Use this state when an interaction can’t occur due to permissions or dependencies.

### Element

![Image Name](assets/3_components/chip/Chips-elements.png)

#### Label (1)

The label of the chip should be short and easily understandable.

#### Checkmark icon (2)

Depending on the type, the close icon or checkmark icon is displayed.

#### Icon (3) (optional)

Add an icon if it makes it easier for users to understand the function of the chip.

## Best Practices

### Chips in groups

- Arrange chips horizontally or vertically.
- Vertical layouts are suitable, for example, for a side menu that should integrate groups of chips.
- Chips arranged horizontally are often located above the content they refer to.
- Word the label as precisely as possible.

## Related components

Tag, Button, Segmented Button, Checkbox Group
