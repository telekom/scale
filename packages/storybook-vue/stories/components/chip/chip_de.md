<div style="display: inline-flex; align-items: center; justify-content: space-between; width: 100%;">
    <h1>Chip</h1>
    <img src="assets/beta.png" alt="Beta Component" />
</div>

Chips sind kontextabhängige Komponenten mit denen Nutzer\*innen Inhalte filtern oder Aktionen auslösen können.

## Allgemein

Chips bestehen aus relevanten Schlüsselwörtern in einem Container. Sie erscheinen häufig dynamisch – ausgelöst durch eine Aktion der Nutzer\*innen – und als Gruppe.

### Verwende einen Chip, um

- Nutzer\*innen Filtermöglichkeiten anzubieten,
- die von Nutzer\*innen gesetzten Filter darzustellen oder
- Nutzer\*innen Optionen vorzuschlagen (Suchkriterien, Antwortmöglichkeiten).

### Verwende Chips besser nicht, wenn

- du ein Element kategorisieren möchtest (hier sind Tags besser geeignet) oder
- wenn du stattdessen einen Button verwenden kannst.

## Basistypen

![Image Name](assets/3_components/chip/Chips-Types.png)

### Persistente Chips

Persistente Chips sind fester Bestandteil des UIs und lassen sich aus- und abwählen.

![Image Name](assets/3_components/chip/Chips-Persistent-Default.png)

**Beispiel 1 – Default** <br/>
Eine Gruppe von Filtermöglichkeiten als fester Bestandteil des UI

![Image Name](assets/3_components/chip/Chips-Persistent-Selected.png)

**Beispiel 2 - Filter ausgewählt** <br/>
Ausgewählte Filter erkennen Nutzer\*innen am Checkmark Icon.

### Dynamische Chips

Dynamische Chips blenden sich nach einer Aktion der Nutzer\*innen ein oder aus.

![Image Name](assets/3_components/chip/Chips-Dynamic-Example1.png)

**Beispiel 1** <br/>
Der Chip blendet sich ein, nachdem Nutzer\*innen einen Filter gesetzt haben. Per Klick/Tap auf das Schließen-Icon blendet er sich wieder aus und der Filter ist zurückgesetzt.

![Image Name](assets/3_components/chip/Chips-Dynamic-Example_Chatbot2.png)

**Beispiel 2** <br/>
Eine Gruppe von Antwortmöglichkeiten in einem Chat. Nach Klick/Tap auf den Chip blendet er sich wieder aus.

## Varianten

![Image Name](assets/3_components/chip/Standard_Chip.png)

Chips lassen sich in zwei Varianten darstellen: Standard und Outline.

### Deaktivierter Zustand

![Image Name](assets/3_components/chip/chips-disabled-de.png)

Im deaktivierten Zustand ist die interaktive Funktion des Chips ausgeschaltet. Verwende diesen Zustand, wenn eine Interaktion aufgrund von Berechtigungen oder Abhängigkeiten nicht möglich ist.

### Elemente

![Image Name](assets/3_components/chip/Chips-elements.png)

#### Label (1)

Das Label des Chips sollte kurz und verständlich sein.

#### Checkmark Icon (2)

Je nach Typ wird das Schließen-Icon oder Checkmark Icon angezeigt.

#### Icon (3) (Optional)

Füge ein Icon hinzu, wenn Nutzer\*innen die Funktion des Chips dadurch leichter verstehen können.

## Best Practice

### Chips in Gruppierungen

- Ordne Chips horizontal oder vertikal an.
- Vertikale Layouts eigenen sich zum Beispiel für ein Seitenmenü, das Chips-Gruppierungen integrieren soll.
- Horizontal angeordnete Chips befinden sich häufig oberhalb vom Inhalt auf den sie sich beziehen.
- Formuliere das Label so präzise wie möglich.

## Beta-Komponente

Diese Komponente befindet sich noch im Beta-Stadium. Wenn du sie testest, bedenke, dass sie möglicherweise noch nicht alle Qualitätskontrollmaßnahmen durchlaufen hat und noch keine WCAG-Zertifizierung zur Barrierefreiheit vorliegt. In Zukunft kann es zu Änderungen an dieser Komponente kommen.

## Verwandte Komponenten

[Tag](?path=/usage/components-tag--standard),
[Button](?path=/usage/components-button--standard),
[Segmented Button](?path=/usage/beta-components-segmented-button--standard),
[Checkbox Group](?path=/usage/components-checkbox-group--standard)
