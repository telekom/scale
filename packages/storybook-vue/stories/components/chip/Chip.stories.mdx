import {
  Meta,
  Story,
  ArgsTable,
  Canvas,
  Description,
} from '@storybook/addon-docs';
import ScaleChip from './ScaleChip.vue';
import { action } from '@storybook/addon-actions';

<Meta
  title="Beta Components/Chip"
  component={ScaleChip}
  argTypes={{
    variant: {
      control: {
        type: 'select',
        options: ['standard', 'outline'],
      },
    },
    type: {
      control: {
        type: 'select',
        options: ['dynamic', 'persistent'],
      },
    },
    disabled: {
      description: `only valid for persistent type`,
    },
  }}
/>

export const Template = (args, { argTypes }) => ({
  props: {
    label: String,
    ...ScaleChip.props,
  },
  template: `
        <scale-chip
            :variant="variant"
            :type="type"
            :selected="selected"
            :disabled="disabled"
            :styles="styles"
        >
        {{label}}
        </scale-chip>
  `,
  methods: {
    action,
  },
});

export const TemplateIcon = (args, { argTypes }) => ({
  props: {
    label: String,
    ...ScaleChip.props,
  },
  template: `
    <scale-chip
      :selected="selected"
      :type="type"
      :label="label"
      :disabled="disabled"
      :styles="styles"
      >
      {{label}}
      <div slot="chip-icon">
        <scale-icon-action-favorite accessibility-title="favorite" />
      </div>
    </scale-chip>
  `,
  methods: {
    action,
  },
});

export const TemplateInteractive = (args, { argTypes }) => ({
  template: `
  <div>
    <div style="height: 20px;">Please select one option:</div>
      <ul aria-label="selection" style="display: flex; list-style: none; padding: 0;">
        <li v-for="label in labels" :key="label" style="margin-right: 4px;">
          <scale-chip
            type="dynamic"
            v-bind:label=label
            @scaleChange="removeChip(label)"
          >
            {{label}}
          </scale-chip>
        </li>
       
      </ul>   
    </div>
  `,
  methods: {
    action: action('scaleChange'),
    removeChip(label) {
      this.action();
      this.labels = null;
      this.text = this.label;
      setTimeout(() => {
        this.$el.querySelector('#root > ul > li > scale-chip') &&
          this.$el
            .querySelector('#root > ul > li > scale-chip')
            .shadowRoot.querySelector('span > button')
            .focus();
      });
    },
  },
  data: () => ({
    labels: ['option 1', 'option 2', 'option 3', 'option 4'],
  }),
});

<div
  style={{
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'space-between',
    width: '100%',
  }}
>
  <h1>Chip</h1>
  <img src="assets/beta.png" alt="Beta Component" />
</div>

## Standard

<Canvas withSource="none">
  <Story
    name="Standard"
    args={{
      label: 'Label',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<ArgsTable story="Standard" />

```html
<scale-chip>{{label}}</scale-chip>
```

## Persistent

<br />

### Persistent Standard

<Canvas withSource="none">
  <Story
    name="Persistent Standard"
    args={{
      label: 'Label',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip>Label</scale-chip>
```

### Persistent Standard Selected

<Canvas withSource="none">
  <Story
    name="Persistent Standard Selected"
    args={{
      label: 'Label',
      selected: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip selected>Label</scale-chip>
```

### Persistent Standard Disabled

<Canvas withSource="none">
  <Story
    name="Persistent Standard Disabled"
    args={{
      label: 'Label',
      disabled: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip disabled selected>Label</scale-chip>
```

### Persistent Standard Selected Disabled

<Canvas withSource="none">
  <Story
    name="Persistent Standard Selected Disabled"
    args={{
      label: 'Label',
      disabled: true,
      selected: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip disabled selected>Label</scale-chip>
```

### Persistent Outline

<Canvas withSource="none">
  <Story
    name="Persistent Outline"
    args={{
      label: 'Label',
      variant: 'outline',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip variant="outline">Label</scale-chip>
```

### Persistent Outline Selected

<Canvas withSource="none">
  <Story
    name="Persistent Outline Selected"
    args={{
      label: 'Label',
      variant: 'outline',
      selected: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip selected>Label</scale-chip>
```

### Persistent Outline Disabled

<Canvas withSource="none">
  <Story
    name="Persistent Outline Disabled"
    args={{
      variant: 'outlined',
      label: 'Label',
      disabled: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip disabled variant="outline" selected>Label</scale-chip>
```

### Persistent Outline Selected Disabled

<Canvas withSource="none">
  <Story
    name="Persistent Outline Selected Disabled"
    args={{
      variant: 'outlined',
      label: 'Label',
      disabled: true,
      selected: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip disabled variant="outline" selected>Label</scale-chip>
```

## Dynamic

### Dynamic Suggestion

<Canvas withSource="none">
  <Story
    name="Dynamic Suggestion"
    args={{
      label: 'Label',
      type: 'dynamic',
      variant: 'standard',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip type="dynamic" variant="standard">Label</scale-chip>
```

### Dynamic Selection Standard

<Canvas withSource="none">
  <Story
    name="Dynamic Selection Standard"
    args={{
      label: 'Label',
      type: 'dynamic',
      variant: 'standard',
      selected: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip type="dynamic" variant="standard" selected>Label</scale-chip>
```

### Dynamic Selection Outline

<Canvas withSource="none">
  <Story
    name="Dynamic Selection Outline"
    args={{
      label: 'Label',
      type: 'dynamic',
      variant: 'outline',
      selected: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-chip type="dynamic" selected variant="outline">Label</scale-chip>
```

## Icon + Text

<Canvas withSource="none">
  <Story
    name="Icon+Text"
    args={{
      label: 'Label',
    }}
  >
    {TemplateIcon.bind({})}
  </Story>
</Canvas>

```html
<scale-chip
  >Label
  <div slot="chip-icon">
    <scale-icon-action-favorite accessibility-title="favorite" />
  </div>
</scale-chip>
```
