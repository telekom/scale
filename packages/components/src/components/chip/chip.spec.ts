import { newSpecPage } from '@stencil/core/testing';
import { Chip } from './chip';

describe('Chip', () => {
  let element;

  beforeEach(async () => {
    element = new Chip();
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Chip],
      html: `<scale-chip>Label</scale-chip>`,
    });
    expect(page.root).toMatchSnapshot();
  });
  it('should handle css classes', () => {
    element.selected = true;
    expect(element.getCssClassMap()).toContain('chip--selected');

    element.disabled = true;
    expect(element.getCssClassMap()).toContain('chip--disabled');
  });
});
