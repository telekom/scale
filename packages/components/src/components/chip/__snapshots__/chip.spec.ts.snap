// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip should match snapshot 1`] = `
<scale-chip>
  <mock:shadow-root>
    <span aria-checked="false" class="chip chip--type-persistent chip--variant-standard" part="base type-persistent variant-standard" role="switch" tabindex="0">
      <slot name="chip-icon"></slot>
      <span class="chip-label">
        <slot></slot>
      </span>
    </span>
  </mock:shadow-root>
  Label
</scale-chip>
`;
